### DISCUSSION TITLE
how an external package C++ code should link against mspass?

### DISCUSSION THREAD

**Discussion by pavlis** [mspass-team/mspass#163] [Design & Development]
As you know I've been trying to adapt the plane wave migration code (pwmig) to be an extension of mspass.   I ran across something that I think is appropriate to put into the discussion pages for mspass because it is worth preserving the solution.

This is partly a cmake confusion but it may have broader implications for extensibility of mspass.   The fundamental issue is I don't understand some magic you are doing in the cmake configurations to handle link libraries.   I observe these things:
1.   The cmake files are compiling all c/c++ code into static libaries (unix ".a" files) found in  cxx/build/src/lib and directories under that one.
2.  The pybind11 binding code compiles are somehow linking against these to build shared libraries (unix .so or mac dylib files).  There is some magic in the CMakeList.txt files there I am not able to fully grok.  

I tried hacking your CMakeList.txt files.  Here is the pattern I borrowed for mspass.seismic:
```
include_directories(
  ${Boost_INCLUDE_DIRS}
  ${GSL_INCLUDE_DIRS}
  ${pylib11_INCLUDE_DIRS}
  ${PYTHON_INCLUDE_DIRS}
  ${PROJECT_SOURCE_DIR}/include
  ${PROJECT_SOURCE_DIR}
  ${PROJECT_BINARY_DIR}/include)

#add_library(mspasspython STATIC ${sources_python})
#target_link_libraries(mspasspython PRIVATE mspass ${PYTHON_LIBRARIES})

pybind11_add_module(pyseismic seismic_py.cc)

target_link_libraries(pyseismic PRIVATE mspass ${Boost_LIBRARIES})
set_target_properties(pyseismic PROPERTIES OUTPUT_NAME "seismic")
if(DEFINED CMAKE_LIBRARY_OUTPUT_DIRECTORY)
  set_target_properties(pyseismic PROPERTIES LIBRARY_OUTPUT_DIRECTORY ${CMAKE_LIBRARY_OUTPUT_DIRECTORY}/ccore)
endif()
```
I changed that for pwmig to the following perturbations:
```
include_directories(
  ${Boost_INCLUDE_DIRS}
  ${GSL_INCLUDE_DIRS}
  ${pylib11_INCLUDE_DIRS}
  ${PYTHON_INCLUDE_DIRS}
  $ENV{MSPASS_HOME}/include
  ${PROJECT_SOURCE_DIR}/include
  ${PROJECT_SOURCE_DIR}
  ${PROJECT_BINARY_DIR}/include)


pybind11_add_module(pyutility utility_py.cc)

target_link_libraries(pyutility PRIVATE $ENV{MSPASS_HOME}/lib/libutility.a ${Boost_LIBRARIES})
set_target_properties(pyutility PROPERTIES OUTPUT_NAME "utility")
if(DEFINED CMAKE_LIBRARY_OUTPUT_DIRECTORY)
  set_target_properties(pyutility PROPERTIES LIBRARY_OUTPUT_DIRECTORY ${CMAKE_LIBRARY_OUTPUT_DIRECTORY}/ccore)
endif()
```
Note that the only difference is changing the name from seismic to utility (for pwmig utility) and insertion in two places of `$ENV{MSPASS_HOME}`.   The include section works BUT the linking does not.  Well, the failure is actually weird.  The linker successfully buiilds an so file but it is invalid.   I try to import it in ipython I get this message:
```
In [1]: import utility
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
<ipython-input-1-6d65e948a93f> in <module>
----> 1 import utility

ImportError: /home/pavlis/src/parallel_pwmig/cxx/build/python/utility.cpython-38-x86_64-linux-gnu.so: undefined symbol: _ZN5pwmig7utility10HypocenterC1Edddd
```
Mysterious until you see this output:
```
objdump -TC utility.cpython-38-x86_64-linux-gnu.so | grep Hypocenter
0000000000000000      D  *UND*	0000000000000000              pwmig::utility::Hypocenter::Hypocenter(double, double, double, double)
0000000000000000      D  *UND*	0000000000000000              pwmig::utility::Hypocenter::Hypocenter(mspass::utility::Metadata const&)
```
The key being "UND".   The two symbols marked that way are those defined by the abbreviated binding code for the Hypocenter object  I'm trying to adapt here.   i.e. these:
```
py::class_<Hypocenter>(m,"Hypocenter","Class to hold earthquake space-time coordinates")
  .def(py::init<>())
  .def(py::init<const mspass::utility::Metadata&>())
  .def(py::init<const double, const double, const double, const double>())
```
I believe that pybind11 has put in a hook for these symbols in the so file but they are not being resolved because when the linker is only defining a hook for these two symbols based on the pybind11 code.  Presumably it thinks dynamic loading will resolve the problem, but it does not.   

That was  very long statement of the problem, but it is complicted.  You guys could help if you could help answer two questions:
1.  How is cmake doing the linking and how can I find the actual commands it uses in mspass and in my clone thereof?  By comparing them I might be able to figure this out.
2. Any idea on what the right incantation I need in CMakeList.txt to address this problem?

A broader question this brings up is if it would be appropriate to build a "lib" directory of pure C/C++ library code without the python bindings?   I suggest `make install` perhaps should put static libraries as it does now but put all those in static.  We could then place dynamic libraries in a "lib" directory.  The pattern I'm thinking is the way the BRTT guys do this in antelope.  They define $ANTELOPE/lib for .so file and $ANTELOPE/static for .a files.   On the negative side this may not be a very high prioritiy as my experience so far is creating a package like this that uses C code im mspass and then build a set of mspass like functionality is an extremely complex procedure it is highly unlikely anyone in our community would take on unless this package became very widely used.

**Comment by wangyinz** [comment]
well, I probably need to look at all the files to get a better understand, but the first thing I would check is to find where the `_ZN5pwmig7utility10HypocenterC1Edddd` symbol is defined. It is very likely that the symbol is also undefined in `$ENV{MSPASS_HOME}/lib/libutility.a`. If so, you will need to link `pyutility` with a library that has that symbol.

**Comment by pavlis** [comment]
Not it is definitely not that the symbol isn't define in the static library.  I'll spare you the details but I double checked with nm --demangle and confirmed the static library defines the symbol.  I am about 99% sure the problem is that the linking procedure is not being pointed at the right directory and it silently creates an invalid .so file because of some trick deep inside pybind11.  That is, I think it is related to the fact that when you try to debug the C code in a python module you always have to create conditional breakpoint.  The C .so files are dynamically loaded only when python asks for them.   This problem is pretty subtle and why I thought it important to preserve it here in github.  

Let me see if I can resolve the linking problem.  If so it will verify my hypothesis.
